libdevcore/Result.h

```diff
@@ -0,0 +1,83 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#pragma once
+
+#include <string>
+
+namespace dev {
+
+/// Simple generic result that holds a value and an optional error message.
+/// Results can be implicitly converted to and created from the type of
+/// the value they hold.
+///
+/// Result<bool> check()
+/// {
+///		if (false)
+///			return Result<bool>(false, "Error message.")
+///		return true;
+/// }
+///
+///	A result can also be instantiated using one of the factory methods it provides:
+///
+/// using BoolResult = Result<bool>;
+/// BoolResult check()
+/// {
+///		if (false)
+///			return BoolResult::failure("Error message");
+///		return BoolResult::success(true);
+/// }
+///
+template <class Type>
+class Result
+{	
+public:
+	using Error = std::string;
+
+	/// @{
+	/// @name Factory functions
+	/// Factory functions that provide a verbose way to create a result
+	static Result<Type> success(Type _value) { return Result(_value); }
+	static Result<Type> failure(Error _error) { return Result(_error); }
+	/// @}
+
+	Result(Type _value): m_value(std::move(_value)) {}
+	Result(Error _error): m_error(std::move(_error)) {}
```

Comment:
You will have problems if ``Type`` is ``std::string`` - do we just want to ignore that?
This would be easy to solve by defining an empty empty ``struct ErrorHelper {};`` and by using that as second argument for the constructor in the error case - especially since you now have the static "construction helpers" above, this could even be private. Or another solution: do we need this constructor anyways or can it just be removed in favor of the one below?